pkg/controller/argocd/util.go

```diff
@@ -1064,10 +1067,98 @@ func containsString(arr []string, s string) bool {
 func namespaceFilterPredicate() predicate.Predicate {
 	return predicate.Funcs{
 		UpdateFunc: func(e event.UpdateEvent) bool {
+			// If ArgoCDManagedByLabel exists, return true.
+			// Event is then handled by the reconciler.
 			if _, ok := e.MetaNew.GetLabels()[common.ArgoCDManagedByLabel]; ok {
 				return true
 			}
+			// This checks if the old meta had the label, if it did, delete the RBACs for the namespace
+			// which were created when the label was added to the namespace.
+			if ns, ok := e.MetaOld.GetLabels()[common.ArgoCDManagedByLabel]; ok && ns != "" {
+				if err := deleteRBACsForNamespace(ns, e.MetaOld.GetName()); err != nil {
+					log.Error(err, fmt.Sprintf("failed to delete RBACs for namespace: %s", e.MetaOld.GetName()))
+				} else {
+					log.Info(fmt.Sprintf("Successfully removed the RBACs for namespace: %s", e.MetaOld.GetName()))
+				}
+			}
 			return false
 		},
 	}
 }
+
+// deleteRBACsForNamespace deletes the RBACs when the label from the namespace is removed.
+func deleteRBACsForNamespace(ownerNS, sourceNS string) error {
+	log.Info(fmt.Sprintf("Removing the RBACs created for the namespace: %s", sourceNS))
+	cfg, err := config.GetConfig()
+	if err != nil {
+		log.Error(err, "unable to get k8s config")
+		return err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		log.Error(err, "unable to create k8s client")
+		return err
+	}
+
+	// List all the roles created for ArgoCD using the label selector
+	labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{common.ArgoCDKeyPartOf: common.ArgoCDAppName}}
+	roles, err := k8sClient.RbacV1().Roles(sourceNS).List(context.TODO(), metav1.ListOptions{LabelSelector: labels.Set(labelSelector.MatchLabels).String()})
+	if err != nil {
+		log.Error(err, fmt.Sprintf("failed to list roles for namespace: %s", sourceNS))
+		return err
+	}
+
+	// Delete all the retrieved roles
+	for _, role := range roles.Items {
```

Comment:
We do set `ownerReferences` to the roles and rolebindings we create, right? I think it would make sense to check whether we are the owner, before deleting the role. I can imagine a scenario where an operator or even a user adds the part-of label with a value of `argocd` manually, and then this role or rolebinding would get removed by accident.

Reply:
Yep, even i had thought of using the `managed-by` label with the value as `argocd-instance-name`, but we don't have the ArgoCD instance details. So i went with `part-of` label instead. 
Let me check if i can somehow retireve the ArgoCD instance name!

Reply:
I think it'd be sufficient for now to check for the `ownerReferences` being an `ArgoCD` resource.

Reply:
As discussed! we can't have a cross-namespace owner reference when the owner itself is namespace scoped. 
So, the resources created by the operator don't have the owner references which could be evaluated. 

Will add a story to track this for the future!

Reply:
Thanks, and you're correct with your finding about cross-namespace ownerReferences. The risk is for edge cases only, so a future implementation is fine for me.